Polyhedron/demo/Polyhedron/Plugins/PMP/Smoothing_plugin.cpp

```diff
@@ -0,0 +1,360 @@
+    Polyhedron& pmesh = (poly_item != NULL) ? * poly_item->polyhedron() : * selection_item->polyhedron();
+
+    int index_id = scene->item_id(poly_item);
+    const double time_step = ui_widget.time_step_spinBox->value();
+    const unsigned int nb_iter = ui_widget.curvature_iter_spinBox->value();
+
+    const bool use_constrained_vertex_map = ui_widget.border_button->isChecked()
+                                            && !CGAL::is_closed(pmesh);
+
+    QApplication::setOverrideCursor(Qt::WaitCursor);
+
+    // explicit scheme
+    if(ui_widget.explicit_checkBox->isChecked())
+    {
+      if(poly_item)
+      {
+        if(use_constrained_vertex_map)
+          smooth_along_curvature_flow(pmesh, time_step, parameters::use_explicit_scheme(true)
+                                                                   .number_of_iterations(nb_iter)
+                                                                   .vertex_is_constrained_map(get_border_constrained_map(pmesh)) );
+        else
+          smooth_along_curvature_flow(pmesh, time_step, parameters::use_explicit_scheme(true)
+                                                                   .number_of_iterations(nb_iter));
+        poly_item->invalidateOpenGLBuffers();
+        poly_item->itemChanged();
+      }
+      else if(selection_item)
+      {
+        if(use_constrained_vertex_map)
+          smooth_along_curvature_flow(selection_item->selected_facets,
+                                      pmesh, time_step, parameters::use_explicit_scheme(true)
+                                                                   .number_of_iterations(nb_iter)
+                                                                   .vertex_is_constrained_map(get_border_constrained_map(pmesh)) );
+        else
+          smooth_along_curvature_flow(selection_item->selected_facets,
+                                      pmesh, time_step, parameters::use_explicit_scheme(true)
+                                                                   .number_of_iterations(nb_iter));
+        selection_item->poly_item_changed();
+        selection_item->changed_with_poly_item();
+      }
+      else
+      {
+        std::cerr<< "Something's gone wrong.\n";
+      }
+    }
+    else // implicit scheme
+    { // Calculate stiffness matrix only once before solving repeatedly
+
+      if(poly_item)
+      {
+        if(index_id != last_index_id || stiffness_is_cleared){
+          // If we changed item or if the stiffness cache is cleared (by the user hitting the button)
+          if(use_constrained_vertex_map){
+            internal::solve_mcf(faces(pmesh), pmesh, time_step, stiffness, true,
+            parameters::vertex_is_constrained_map(get_border_constrained_map(pmesh)) );
+          }
+          else{
+            internal::solve_mcf(faces(pmesh), pmesh, time_step, stiffness, true,
+            parameters::all_default());
+          }
+          last_index_id = index_id;
```

Comment:
this mechanism to use some caching for an item is buggy as it will not take into account the reloading of an item for example